### PR TITLE
YlmSpherepack: Replace mutable non-dynamic storage

### PR DESCRIPTION
**Pull request by markscheel**
## Proposed changes

This addresses part of #3929 but not all of it.
    
YlmSpherepack has 2 types of internal storage:
1. Quantities computed only once and reused often, but forever constant.
 2. Temporary buffers that are overwritten frequently.
    
 This PR changes 1. and does not address 2.
    
Previously, the temporary quantities in 1. were computed using
lazy evaluation: if a member function that needed them was
called, then they were computed and then forever held fixed.
This was accomplished via a mutable member variable that holds the
storage.  Now, the temporary quantities in 1. are all computed
in the YlmSpherepack constructor, and are never changed.  The
corresponding member variable is no longer mutable.  In the
constructor, certain temporary quantities are allocated on the heap
on the fly and then thrown away after use, but because this happens
only in the constructor we don't worry about these allocations.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp
Comment (nilsdeppe) [inline src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp]: Do you want to mark these checks `UNILKELY`? I'm fine either way :)

**Review by nilsdeppe — DISMISSED** [review]
LGTM!

**Review by nilsdeppe — DISMISSED** [review]
LGTM, if you rebase on develop the tests should work again :)

**Comment by markscheel** [comment]
Squashed and rebased.

**Review by nilsdeppe — DISMISSED** [review]

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp
Comment (nilsvu) [inline src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp:55]: Commit message: typo in 2. (delete "call"?)

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp
Comment (nilsvu) [inline src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp:684]: Can you just use `std::vector` to allocate some memory? No reason to use `DataVector` just for allocating memory I think

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.hpp
Comment (nilsvu) [inline src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.hpp:30]: Would it be reasonable to make these members either `DataVector`s or [`blaze::DynamicVector`](https://bitbucket.org/blaze-lib/blaze/wiki/Vector%20Types)s, so arithmetic operations can run through Blaze? That would eliminate many of the loops in the code I think. My understanding is that a nonowning DataVector is pretty much a gsl::span. There's also the Blaze-native [`blaze::CustomVector`](https://bitbucket.org/blaze-lib/blaze/wiki/Vector%20Types#!customvector), which is essentially a gsl::span.

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.hpp
Comment (nilsvu) [inline src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.hpp:452]: I got confused that this is named `ConstStorage` but is not actually const, and many member functions retrieve non-const references into it. Can you make this member variable `const`? If needed, maybe move the functions that fill it (such as `calculate_collocation_points`) to `YlmSpherepack_detail` and call them in the constructor of `ConstStorage`?

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp
Comment (nilsdeppe) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp:684]: `DataVector` is better than `std::vector<double>`. It default initializes to `std::signaling_NaN()` in debug (helps catch errors) and doesn't default initialize in release (more efficient). In general, I think we should use fewer `std::vector<double>` and more `DataVector`. Note that Blaze's vectors would not be a suitable replacement for `DataVector` since we have no guarantees on how they initialize memory.

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.hpp
Comment (nilsdeppe) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.hpp:30]: As in my other comment `DataVector` should pretty much always be preferred over `std::vector<double>` and `blaze::DynamicVector<double>`

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp
Comment (nilsvu) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp:684]: Ok. Tangent unrelated to this PR: my concern with overusing DataVector is that it has NodalVector semantics in many places, meaning that it represents data at grid points and supports associated arithmetics, as opposed to ModalVector that explicitly forbids some operations. Perhaps we should keep DataVector as a general-purpose dense vector with the memory allocation behavior that you mentioned, and add a new NodalVector type specifically for data at grid points. What do you think?

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.hpp
Comment (nilsvu) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.hpp:30]: Ok, DataVector then :)

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp
Comment (nilsdeppe) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp:684]: I have no objection to doing such a rename and adding some `TempVector<T>` class that behaves like `DataVector` in terms of allocations and non-owning support. We should assess the amount of work this would be, both for changing and reviewing a PR, that's my only concern. In any case, I think that should be done in a separate PR :)

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.hpp
Comment (markscheel) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.hpp:452]: Can I instead rename `ConstStorage` to `Storage`?

I tried making the member variable `const`, and I am going down a rabbit hole of defining custom copy-assignment operators (and copy constructors, move constructors, etc) for many classes.  Apparently `const` member functions don't play well with copy assignment.  And then I saw this: https://stackoverflow.com/questions/4136156

The only member functions that should get non-const refs to `storage_` are the non-const member functions.

Comment (nilsdeppe) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.hpp:452]: In general, I've found it's a huge pain to have `const` member variables, as you stated. I don't see an issue with the current design, to be honest. I'd be fine with naming it to `const_storage_` as the member variable too... I don't think this is generally worth the effort. The design is correct, the comments I think are clear

**Comment by markscheel** [comment]
I closed this previously by mistake.

Now rebased onto develop, with conflicts fixed.

**Review by nilsdeppe — DISMISSED** [review]

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp
Comment (markscheel) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.cpp:55]: fixed.

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.hpp
Comment (nilsvu) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.hpp:452]: Ok, perhaps make the refs into storage_ const in the member functions, meaning `auto& cos_theta = ...` -> `const auto& cos_theta = ...`?

Comment (nilsdeppe) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.hpp:452]: Seems reasonable, but if the member functions are `const` (I haven't check if they all are) then `const auto&` and `auto&` would be equivalent (except for readability)

**Comment by markscheel** [comment]
Changed `auto&` to `const auto&` when accessing parts of `_storage`.  And squashed.

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
@nilsvu could you please take a look at this again? Thanks!

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.hpp
Comment (nilsvu) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.hpp:30]: Opinions on this?

**Comment by nilsvu** [comment]
Opinions on the thread above https://github.com/sxs-collaboration/spectre/pull/3931#discussion_r856894221?

Please don't wait for my review on this PR, I don't have enough time right now to do a proper review, so I'll just remove myself as a reviewer. Sorry!

File: src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.hpp
Comment (markscheel) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.hpp:30]: This would be an interface change.   If people want this, could it be done in a separate PR?

Comment (nilsdeppe) [reply src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.hpp:30]: I wouldn't worry about this now. It's the sort of thing where if we profile and it shows up we can change it. When adding code it's I think good to be aware of the performance implications, but I agree it's beyond the scope of this PR.

**Pull request merged**
